extensions/src/data_bridge/index.ts

```diff
@@ -17,75 +19,107 @@ const plugin: JupyterFrontEndPlugin<void> = {
+                        // @ts-ignore
+                        kernel.dataFromHost = app.dataFromHost;
+                        loadData(kernel, app.dataFromHost);
+                    }
+                    // Reset the flag when the kernel is restarting, since this flag is not affected by the kernel restart
+                    if (status === 'restarting') {
+                        kernel.dataFromHost = "";
+                    }
+                });
             }
         });
 
+        /**
+         * Send data to the host page
+         * @param data
+         */
         // @ts-ignore
-        window.sendDataToHost = (data: any) => {
+        window.sendDataToHost = (data: object) => {
             window.parent.postMessage(
                 {
                     type: "from-iframe-to-host",
-                    data: data,
+                    action: "set-data",
+                    payload: data
                 },
                 "*"
             );
         };
 
+        /**
+         * Listen for messages from the host page, and update the data in the kernel
+         * @param event MessageEvent
+         */
         // @ts-ignore
-        window.requestDataFromHost = (variableName = "data") => {
-            window.parent.postMessage(
-                {
-                    type: "from-iframe-to-host",
-                    requestData: true,
-                    variableName,
-                },
-                "*"
-            );
-        };
-
-        window.addEventListener("message", async (event) => {
+        window.addEventListener("message", async (event: MessageEvent<IframeMessageSchema>) => {
             if (event.data.type === "from-host-to-iframe") {
-                let data = event.data.data;
-                let variableName = event.data.variableName || "data";
-                const dataJson = JSON.stringify(data);
-                const code = `
-  import json
-  ${variableName} = json.loads('${dataJson}')
-      `;
-                // Similar to https://jupyterlab.readthedocs.io/en/stable/api/classes/application.LabShell.html#currentWidget
-                // https://jupyterlite.readthedocs.io/en/latest/reference/api/ts/interfaces/jupyterlite_application.ISingleWidgetShell.html#currentwidget
-                const currentWidget = app.shell.currentWidget;
-
-                if (currentWidget instanceof NotebookPanel) {
-                    const notebookPanel = currentWidget;
-                    const kernel = notebookPanel.sessionContext.session?.kernel;
-                    if (kernel) {
-                        kernel.requestExecute({ code: code });
-                    } else {
-                        console.error("No active kernel found");
-                    }
-                } else {
-                    console.error("Current active widget is not a notebook");
-                }
+                // @ts-ignore
+                app.dataFromHost = JSON.stringify(event.data.payload);
```

Comment:
JSON stringify should be executed before we send the message, probably and/or we should do the validation with schema

Reply:
Well, validation happens when we do the `entity.toJSON()` and send it, then here we use stringify due to how it's going to be consumed by pyodide code execution.